### PR TITLE
feat: Add slotFormat query parameter to slots API V2 endpoint to return start and end time of a slot

### LINKED ISSUE #10548: Slots Endpoint does not Provide End Time for each Slot
### Is your proposal related to a problem?
This proposal is to make the slots endpoint easier to use to generate an array of available time slots with start AND end time in the response. As a Cal.com user that only uses the backend functionality and does not use the front-end functionality, the seamlessness of the API is crucial for us.

Currently, the Slots endpoint within the Cal.com API does not provide an end time for each slot. Although one may implicitly determine it by checking the duration assigned to the eventTypeId, it does not make the response very user friendly. It seems counterintuitive to then attempt to use the create booking endpoint as it requires a start and end time in the request body.

# Example Response:
```json
{
    "slots": {
        "2023-08-03": [
            {
                "time": "2023-08-03T13:00:00.000Z",
                "users": []
            },
            {
                "time": "2023-08-03T13:30:00.000Z",
                "users": []
            },
            {
                "time": "2023-08-03T14:00:00.000Z",
                "users": []
            },
            {
                "time": "2023-08-03T14:30:00.000Z",
                "users": []
            },
            {
                "time": "2023-08-03T15:00:00.000Z",
                "users": []
            },
            {
                "time": "2023-08-03T15:30:00.000Z",
                "users": []
            },
            {
                "time": "2023-08-03T16:00:00.000Z",
                "users": []
            },
            {
                "time": "2023-08-03T16:30:00.000Z",
                "users": []
            },
            {
                "time": "2023-08-03T17:00:00.000Z",
                "users": []
            },
            {
                "time": "2023-08-03T17:30:00.000Z",
                "users": []
            }
        ]
    }
}
```
### Describe the solution you'd like

The proposed improvement would make the current response to the API (example above) look more like this:
```json
{
    "slots": {
        "2023-08-03": [
            {
                "startTime": "2023-08-03T13:00:00.000Z",
                "endTime": "2023-08-03T13:30:00.000Z",
                "users": []
            },
            {
                "startTime": "2023-08-03T13:30:00.000Z",
                "endTime": "2023-08-03T14:00:00.000Z",
                "users": []
            },
            {
                "startTime": "2023-08-03T14:00:00.000Z",
                "endTime": "2023-08-03T14:30:00.000Z",
                "users": []
            },
            {
                "startTime": "2023-08-03T14:30:00.000Z",
                "endTime": "2023-08-03T15:00:00.000Z",
                "users": []
            },
            {
                "startTime": "2023-08-03T15:00:00.000Z",
                "endTime": "2023-08-03T15:30:00.000Z",
                "users": []
            },
            {
                "startTime": "2023-08-03T15:30:00.000Z",
                "endTime": "2023-08-03T16:00:00.000Z",
                "users": []
            },
            {
                "startTime": "2023-08-03T16:00:00.000Z",
                "endTime": "2023-08-03T16:30:00.000Z",
                "users": []
            },
            {
                "startTime": "2023-08-03T16:30:00.000Z",
                "endTime": "2023-08-03T17:00:00.000Z",
                "users": []
            },
            {
                "startTime": "2023-08-03T17:00:00.000Z",
                "endTime": "2023-08-03T17:30:00.000Z",
                "users": []
            },
            {
                "startTime": "2023-08-03T17:30:00.000Z",
                "endTime": "2023-08-03T18:00:00.000Z",
                "users": []
            }
        ]
    }
}
```
### Describe alternatives you've considered

1. I have considered an assumption of duration based on a check of details of the eventTypeId
2. I have considered programmatically determining the endTime on the front-end for each slot

### PR DESCRIPTION
**Pull request by asadath1395**
## What does this PR do?

Add slotFormat query parameter to slots API V2 endpoint to return start and end time of a slot

- Fixes #10548

https://www.loom.com/share/3a3d4d3035294955bb7c1a778d1a22ea?sid=fc036065-06d5-402f-ad41-88426272a930

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Create a new event type
- Using the event type details from above step, hit `/v2/slots/available` endpoint

### CODE REVIEW

**Review by Praashh — CHANGES_REQUESTED** [review]
Would be great, if you can add a loom video @asadath1395

**Comment by asadath1395** [comment]
@Praashh I have added a loom video

File: apps/api/v2/src/modules/slots/controllers/slots.controller.ts [hunk 1]
```diff
@@ -129,6 +158,44 @@ export class SlotsController {
...
+      } else {
+        throw new Error("duration or eventTypeId is required");
+      }
+
+      const transformedSlots = Object.entries(availableSlots.slots).reduce<
+        Record<string, { startTime: string; endTime: string }[]>
+      >((acc, [date, slots]) => {
+        acc[date] = (slots as { time: string }[]).map((slot) => {
+          const startTime = new Date(slot.time);
+          const endTime = new Date(startTime.getTime() + duration * 60000); // Convert minutes to milliseconds
```
Comment (Praashh) [inline apps/api/v2/src/modules/slots/controllers/slots.controller.ts:182]:
```suggestion
          const endTime = new Date(startTime.getTime() + duration * 60000); 
```

**Comment by asadath1395** [comment]
@supalarry or @ThyMinimalDev Can any of you review this PR whenever you have time? Thanks!

File: apps/api/v2/src/modules/slots/controllers/slots.controller.ts [hunk 2]
```diff
@@ -129,6 +158,44 @@ export class SlotsController {
...
+          throw new Error("Event type not found");
+        }
+
+        duration = eventType.length;
+      } else {
+        throw new Error("duration or eventTypeId is required");
+      }
+
+      const transformedSlots = Object.entries(availableSlots.slots).reduce<
+        Record<string, { startTime: string; endTime: string }[]>
```
Comment (Praashh) [inline apps/api/v2/src/modules/slots/controllers/slots.controller.ts:178]: we need to make sure availableSlots.slots is always defined and is an object. If not, the Object.entries(availableSlots.slots) call may throw an error.

```suggestion
     if (!availableSlots || !availableSlots.slots) {
         throw new Error("Available slots are missing"); 
      }
      const transformedSlots = Object.entries(availableSlots.slots).reduce<
        Record<string, { startTime: string; endTime: string }[]>
```

File: apps/api/v2/src/modules/slots/controllers/slots.controller.ts [hunk 3]
```diff
@@ -129,6 +158,44 @@ export class SlotsController {
...
+      >((acc, [date, slots]) => {
+        acc[date] = (slots as { time: string }[]).map((slot) => {
+          const startTime = new Date(slot.time);
+          const endTime = new Date(startTime.getTime() + duration * 60000);
```
Comment (Praashh) [inline apps/api/v2/src/modules/slots/controllers/slots.controller.ts:182]: make sure the duration is always a valid positive

File: apps/api/v2/src/modules/slots/controllers/slots.controller.ts (hunk 2 above)
Comment (asadath1395) [reply apps/api/v2/src/modules/slots/controllers/slots.controller.ts:178]: It is already being handled by `getAvailableSlots`, it always returns slots, ref: https://github.com/calcom/cal.com/blob/4eac6c755dace7ae2c742c0d5e3a88283acd3942/packages/trpc/server/routers/viewer/slots/util.ts#L790

File: apps/api/v2/src/modules/slots/controllers/slots.controller.ts [hunk 4]
```diff
@@ -129,6 +158,44 @@ export class SlotsController {
...
+        duration = eventType.length;
+      } else {
+        throw new Error("duration or eventTypeId is required");
+      }
+
+      const transformedSlots = Object.entries(availableSlots.slots).reduce<
+        Record<string, { startTime: string; endTime: string }[]>
+      >((acc, [date, slots]) => {
+        acc[date] = (slots as { time: string }[]).map((slot) => {
+          const startTime = new Date(slot.time);
```
[the remark this answers is not in the document]
Comment (asadath1395) [reply apps/api/v2/src/modules/slots/controllers/slots.controller.ts:181]: We are calling already working function `getAvailableSlots` which is used by the slots API in UI also

File: apps/api/v2/src/modules/slots/controllers/slots.controller.ts (hunk 3 above)
Comment (asadath1395) [reply apps/api/v2/src/modules/slots/controllers/slots.controller.ts:182]: Fixed

**Comment by asadath1395** [comment]
Hey @Praashh 
Thanks for taking the time to review the PR, i have added comments/fixes wherever applicable. Do check that and out and let me know

**Review by Praashh — DISMISSED** [review]
LGTM!!

File: apps/api/v2/swagger/documentation.json [hunk 5]
```diff
@@ -5003,6 +5003,16 @@
             "schema": {
               "type": "number"
             }
+          },
+          {
+            "name": "formatAsStartAndEndTime",
```
Comment (ThyMinimalDev) [inline apps/api/v2/swagger/documentation.json:5008]: formatAsStartAndEndTime seems overly verbose, I would prefer something like slotFormat=range

File: apps/api/v2/src/modules/slots/controllers/slots.controller.ts [hunk 6]
```diff
@@ -129,6 +158,44 @@ export class SlotsController {
...
+          const endTime = new Date(startTime.getTime() + duration * 60000); 
+          return {
+            startTime: startTime.toISOString(),
+            endTime: endTime.toISOString(),
+          };
+        });
+        return acc;
+      }, {});
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/slots/controllers/slots.controller.ts:189]: place this whole function in slots.service, name as formatSlots for example, have it return the formatted slots

**Review by ThyMinimalDev — CHANGES_REQUESTED** [review]

**Comment by asadath1395** [comment]
Hey @ThyMinimalDev 

Thanks for reviewing this PR. I have made the changes as you suggested, please do review this when you get time. Thanks!

File: packages/platform/types/slots.ts [hunk 7]
```diff
@@ -57,6 +70,22 @@ export class GetAvailableSlotsInput {
...
+  @IsString()
+  @IsEnum(SlotFormat, {
+    message: "slotFormat must be either 'range' or 'time'",
+  })
+  @Transform(({ value }) => {
+    if (!value) return undefined;
+    return value.toLowerCase();
+  })
+  @IsOptional()
+  @ApiProperty({
```
Comment (ThyMinimalDev) [inline packages/platform/types/slots.ts:83]: ApiPropertyOptional

Comment (asadath1395) [reply packages/platform/types/slots.ts]: Fixed

**Review by ThyMinimalDev — APPROVED** [review]

**Pull request merged**